// @xmtp/content-type-text/dist/index.cjs
'use strict';

var contentTypePrimitives = require('@xmtp/content-type-primitives');

const ContentTypeText = new contentTypePrimitives.ContentTypeId({
    authorityId: "xmtp.org",
    typeId: "text",
    versionMajor: 1,
    versionMinor: 0,
});
exports.Encoding = void 0;
(function (Encoding) {
    Encoding["utf8"] = "UTF-8";
})(exports.Encoding || (exports.Encoding = {}));
class TextCodec {
    get contentType() {
        return ContentTypeText;
    }
    encode(content) {
        return {
            type: ContentTypeText,
            parameters: { encoding: exports.Encoding.utf8 },
            content: new TextEncoder().encode(content),
        };
    }
    decode(content) {
        const { encoding } = content.parameters;
        if (encoding !== exports.Encoding.utf8) {
            throw new Error(`unrecognized encoding ${encoding}`);
        }
        return new TextDecoder().decode(content.content);
    }
    fallback() {
        return undefined;
    }
    shouldPush() {
        return true;
    }
}

exports.ContentTypeText = ContentTypeText;
exports.TextCodec = TextCodec;
